### PR TITLE
[Whisper ASR] Remove decode-path GPU sync and vectorize logit_bias build, raise pre_lm_max_batch_size default

### PR DESCRIPTION
**Pull request by Dayuxiaoshui**
## Motivation

`py-spy` profiling of Whisper ASR's `scheduler-asr` thread at concurrency 16 found that ~32% of the thread's sampled time was non-model-forward overhead concentrated in two upstream-sglang hotspots hit on every decode/extend step for encoder-decoder models:

- `DecodeCudaGraphRunner.can_run_graph` / `_can_run_ragged_verify_graph` compute `torch.all(forward_batch.encoder_lens > 0)` inside a Python `if`/`and` expression. The surrounding `bool()` coercion forces a CUDA device synchronize + device-to-host copy on every decode step, even though `ForwardBatch` already carries an equivalent CPU-side mirror (`encoder_lens_cpu`) that was never used here. This alone was ~20.5% of the thread's sampled time.
- `SamplingBatchInfo.from_schedule_batch` builds the dense `logit_bias` tensor with a nested Python loop — one `logit_bias[i, int(key)] = value` CUDA kernel launch per suppressed token per request. Whisper applies `generation_config.suppress_tokens` via `logit_bias` (~90 keys), and every request repeats the same write set on every extend/prefill batch-composition step. This was ~11.1% of the thread's sampled time.

Separately, `pre_lm_max_batch_size` (the cap on Whisper's pre-LM encoder CUDA graph batch size) defaulted to 8, independent of `max_running_requests=64`, forcing encoder batches above 8 into eager fallback even though the default graph bucket list (`(1, 2, 4, 8, 12, 16)`) already covers 16 with no code change needed.

This PR fixes both CPU-overhead sources with omni-side monkeypatches (following the existing `sglang_omni/vendor/sglang/` convention already used for `apply_qk_norm` and `RMSNorm.forward_cuda`) rather than an upstream sglang PR, and raises the `pre_lm_max_batch_size` default from 8 to 16. None of this changes model architecture, attention behavior, KV cache, sampling semantics, or checkpoint format — only how two existing computations are executed and one existing batch-size knob's default value.

## Modifications

- Add `sglang_omni/vendor/sglang/model_executor.py`: patches `DecodeCudaGraphRunner.can_run_graph` and `_can_run_ragged_verify_graph` to compute the encoder-lens-supported predicate from `ForwardBatch.encoder_lens_cpu` (a CPU list already populated by sglang) with no GPU sync, falling back to the original tensor-based check whenever the CPU mirror is `None` so behavior for any caller that doesn't populate it is unchanged. General fix, not Whisper-specific.
- Add `sglang_omni/vendor/sglang/sampling.py`: patches `SamplingBatchInfo.from_schedule_batch` to collect all `(row, col, value)` `logit_bias` triples on the CPU first, then materialize the tensor with a single `torch.zeros(...)` + `index_put_(...)` call instead of one kernel launch per key per request. The rest of the classmethod (temperature/top-p/top-k/penalizer/custom-logit-processor construction) is unchanged.
- Wire both patches in via side-effect imports (`# noqa: F401`) in `sglang_omni/models/whisper_asr/engine_builder.py`, scoped to Whisper since that's the validated context — the fixes fall back to unchanged behavior for other omni ASR models (`is_encoder_decoder=False`, `encoder_lens_cpu=None`) but weren't separately benchmarked there.
- Raise `pre_lm_max_batch_size` default 8 → 16 in `engine_builder.py`, `stages.py`, and `config.py`, with matching test/doc updates (`tests/unit_test/whisper_asr/test_pipeline.py`, `docs/cookbook/whisper_asr.md`).

Both new vendor files carry a module docstring citing the exact upstream file/line and the sglang version they were captured against (`sglang==0.5.16`), with a re-diff warning for future upgrades — they're full method overrides rather than call-through wrappers because both hot spots sit mid-expression inside a larger, non-separable method.

## Accuracy Test

Setup: `openai/whisper-large-v3` at revision `06f233fe06e710322aca913c1bc4249a0d71fce1`, 200-sample SeedTTS English subset at revision `27f4c1adee83b5b29b7c4b375f6b976324bda308`, one NVIDIA H200, TP=1, FP16, greedy decoding, generation and pre-LM encoder CUDA graphs enabled, `pre_lm_max_batch_size=16` in both arms.

| Concurrency | Patches ON corpus WER | Patches OFF corpus WER |
|---:|---:|---:|
| 16 | 0.0097 | 0.0097 |
| 32 | 0.0097 | 0.0097 |

Corpus WER is identical with the patches on and off at every measured concurrency — both patches are behavior-preserving by construction (one substitutes an equivalent CPU-computed predicate with a tensor-check fallback, the other substitutes an equivalent vectorized tensor-write), and this holds up empirically. (The 0.0097 value reflects this 200-sample subset, not the 1,088-clip full-corpus number elsewhere in the Whisper WER investigation — not a regression, see `docs/whisper_wer_gap_and_profiling.md` §3.2.)

Additional correctness checks (not full accuracy runs): `_is_encoder_lens_supported` was exercised against all-positive / one-zero / GPU-tensor-fallback / non-encoder-decoder cases; `_build_logit_bias`'s output was checked bit-for-bit (`torch.equal`) against the original loop's output across a multi-request example including a `None`-bias request. The full `tests/unit_test` suite passes unchanged (pre-existing, unrelated `dots_tts` collection error aside).

## Benchmark & Profiling

Same setup as the accuracy test. **Isolated A/B** — same server config (`pre_lm_max_batch_size=16`), same GPU, same port, back-to-back (patches enabled, then the two vendor imports commented out and the server restarted, then restored), 3 repeats per concurrency, values below are the mean across repeats:

| Concurrency | Patches ON req/s | Patches OFF req/s | Throughput Δ | Patches ON mean latency (s) | Patches OFF mean latency (s) | Mean latency Δ | Patches ON P95 latency (s) | Patches OFF P95 latency (s) | P95 latency Δ | Patches ON P99 latency (s) | Patches OFF P99 latency (s) | P99 latency Δ |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 16 | 65.46 | 59.62 | **+9.8%** | 0.241 | 0.265 | **-8.9%** | 0.352 | 0.375 | **-5.9%** | 0.396 | 0.430 | **-7.9%** |
| 32 | 88.49 | 80.45 | **+10.0%** | 0.354 | 0.390 | **-9.2%** | 0.532 | 0.610 | **-12.7%** | 0.640 | 0.713 | **-10.3%** |

Corpus WER was identical between arms at both concurrencies (see the Accuracy Test table above), so every one of these gains is throughput/latency-only with no accuracy trade-off. `py-spy` on the `scheduler-asr` thread corroborates the mechanism: `can_run_graph` no longer appears anywhere in sampled stack traces (previously ~20.5%), and `from_schedule_batch` drops to ~2.1% of samples (previously ~11.1%). No crashes, no eager-mode fallback warnings, and no `torch.equal` mismatches were observed in any run.

Separately, raising `pre_lm_max_batch_size` from 8 to 16 alone (no patches) was benchmarked back-to-back against the old default of 8 on the same subset:

| Concurrency | `pre_lm_max_batch_size=8` (req/s mean) | `pre_lm_max_batch_size=16` (req/s mean) | Δ |
|---:|---:|---:|---:|
| 16 | ~58.9 | ~61.9 | **+~5%** |
| 32 | ~77.7 | ~78.7 | ~+1% (noise-level) |

An earlier non-isolated comparison (patches run once, compared cross-session against a `pre_lm_max_batch_size=16`-only baseline) had shown an apparent throughput dip at concurrency 16; the isolated A/B above resolves that as cross-run/GPU-contention noise, not a real regression — see `docs/whisper_wer_gap_and_profiling.md` §3.3 for the full writeup.

## Checklist

- [x] Format your code according with pre-commit. (`isort`/`black` clean on all changed files.)
- [ ] Add unit tests. `test_pipeline.py` covers the `pre_lm_max_batch_size` default change; the two new vendor patches were validated with ad-hoc correctness checks (`torch.equal`, edge-case exercising — see Accuracy Test above) but do not yet have committed pytest cases. Worth adding as a follow-up.
- [x] Update documentation / docstrings / example tutorials as needed. (`docs/cookbook/whisper_asr.md` updated for the new default; both vendor patch files document their upstream source and rationale.)
- [x] Provide throughput / latency benchmark results and accuracy evaluation results as needed.
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.

## CI

CI runs on self-hosted GPU runners and requires a maintainer to add the `run-ci` label. Once labeled, every subsequent push re-triggers CI as long as the label remains. Use `/tag-and-rerun-ci higgs` or `/tag-and-rerun-ci moss` or `/tag-and-rerun-ci qwen3-tts` to select a TTS CI model, and `/tag-and-rerun-ci fun-asr`, `/tag-and-rerun-ci qwen3-asr` or `/tag-and-rerun-ci whisper-asr` to select an ASR CI model. One selector from each family can be combined, for example `/tag-and-rerun-ci moss fun-asr`. Draft PRs are skipped even if labeled.

### CODE REVIEW

File: sglang_omni/models/whisper_asr/engine_builder.py [hunk 1]
```diff
@@ -237,6 +249,22 @@ def adjust_overrides(self, overrides: dict[str, Any]) -> None:
         # per-request processor; this flag permits SGLang to execute it.
         overrides["enable_custom_logit_processor"] = True
 
+        attention_backend = overrides.get("attention_backend")
```
Comment (AkazaAkane) [inline sglang_omni/models/whisper_asr/engine_builder.py]: This checks only the base attention backend, not the effective decode backend. For example, `attention_backend="flashinfer"` with `decode_attention_backend="fa3"` still reaches the known crashing decode-CUDA-graph path, while the inverse configuration disables graphs unnecessarily.

Also, setting global `disable_cuda_graph=True` disables Whisper's encoder CUDA Graph even though the incompatibility described here is decode-only. Please resolve the effective decode backend and set `disable_decode_cuda_graph=True`, or reject the unsafe combination explicitly.

Please add tests for split prefill/decode backends and explicit/nested decode CUDA Graph configuration.

File: sglang_omni/models/whisper_asr/stages.py [hunk 2]
```diff
@@ -28,7 +28,7 @@ def create_sglang_whisper_asr_executor(
     enable_pre_lm_encoder: bool = True,
     pre_lm_cache_max_entries: int = 1024,
     pre_lm_cache_size_bytes: int | None = None,
-    pre_lm_max_batch_size: int = 8,
+    pre_lm_max_batch_size: int = 16,
```
Comment (AkazaAkane) [inline sglang_omni/models/whisper_asr/stages.py:31]: The default `request_build_max_workers` is still 8, and each cache-miss request builder blocks in `encode_item()` until its encoder future completes. That appears to limit the number of outstanding encoder jobs to 8, so encoder batches 12/16 cannot occur under the default configuration. For TP > 1, the scheduler reduces request-build workers to 1, making the reachable batch size even smaller.

Despite that, the 16/12 graph buckets are captured first and retain graph inputs/outputs, potentially consuming VRAM before the actually reachable buckets are captured. Please cap encoder graph buckets by the effective request-build concurrency, or provide the realized encoder batch-size histogram and post-capture VRAM evidence showing that batches 12/16 replay under the shipped defaults.

File: sglang_omni/vendor/sglang/sampling.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+            rows.append(i)
+            cols.append(int(key))
+            values.append(float(value))
+
+    logit_bias = torch.zeros(len(reqs), vocab_size, device=device)
+    if rows:
+        row_idx = torch.tensor(rows, dtype=torch.long, device=device)
+        col_idx = torch.tensor(cols, dtype=torch.long, device=device)
+        val = torch.tensor(values, dtype=logit_bias.dtype, device=device)
+        logit_bias.index_put_((row_idx, col_idx), val)
```
Comment (AkazaAkane) [inline sglang_omni/vendor/sglang/sampling.py:71]: This is not fully equivalent to the original loop when distinct string keys canonicalize to the same token ID. The original loop has deterministic last-write-wins behavior, while `index_put_` with duplicate indices and `accumulate=False` does not guarantee that result. Please canonicalize `(row, int(key)) -> value` on the CPU before creating the index tensors, preserving insertion-order last-write-wins semantics, and add an old-vs-new parity test for duplicate canonical token IDs.

**Review by AkazaAkane — CHANGES_REQUESTED** [review]
The profiling result and isolated A/B are convincing. I am requesting changes for three below issues.

There is also a maintenance risk from replacing complete SGLang 0.5.16 methods. The pending 0.5.18 upgrade can change these methods without producing a merge conflict, after which these monkeypatches would silently restore stale 0.5.16 behavior.
